src/encoder/mod.rs

```diff
@@ -0,0 +1,494 @@
+    }
+
+    /// Write some data to the tiff file, the offset of the data is returned.
+    ///
+    /// This could be used to write tiff strips.
+    pub fn write_data<T: TiffValue>(&mut self, value: T) -> TiffResult<u64> {
+        let offset = self.writer.offset();
+        value.write(&mut self.writer)?;
+        Ok(offset)
+    }
+
+    fn finish_internal(&mut self) -> TiffResult<()> {
+        let ifd_pointer = self.write_directory()?;
+        let curr_pos = self.writer.offset();
+
+        self.writer.goto_offset(self.ifd_pointer_pos)?;
+        self.writer.write_u32(ifd_pointer as u32)?;
+        self.writer.goto_offset(curr_pos)?;
+        self.writer.write_u32(0)?;
+
+        self.dropped = true;
+
+        Ok(())
+    }
+
+    /// Write out the ifd directory.
+    pub fn finish(mut self) -> TiffResult<()> {
+        self.finish_internal()
+    }
+}
+
+impl<'a, W: Write + Seek> Drop for DirectoryEncoder<'a, W> {
+    fn drop(&mut self) {
+        if !self.dropped {
+            let _ = self.finish_internal();
+        }
+    }
+}
+
+
+/// Type to encode images strip by strip.
+///
+/// You should call `finish` on this when you are finished with it.
+/// Encoding can silently fail while this is dropping.
+///
+/// # Examples
+/// ```
+/// # extern crate tempfile;
+/// # let mut file = tempfile::tempfile().unwrap();
+/// # let image_data = vec![0; 100*100*3];
+/// use tiff::encoder::*;
+///
+/// let mut tiff = TiffEncoder::new(&mut file).unwrap();
+/// let mut image = tiff.new_image::<colortype::RGB8>(100, 100).unwrap();
+///
+/// let mut idx = 0;
+/// while image.next_strip_sample_count() > 0 {
+///     let sample_count = image.next_strip_sample_count() as usize;
+///     image.write_strip(&image_data[idx..idx+sample_count]).unwrap();
+///     idx += sample_count;
+/// }
+/// image.finish().unwrap();
+/// ```
+pub struct ImageEncoder<'a, W: Write + Seek, C: ColorType> {
+    encoder: DirectoryEncoder<'a, W>,
+    strip_idx: u64,
+    strip_count: u64,
+    row_samples: u64,
+    height: u32,
+    rows_per_strip: u64,
+    strip_offsets: Vec<u32>,
+    strip_byte_count: Vec<u32>,
+    dropped: bool,
+    _phantom: std::marker::PhantomData<C>,
+}
+
+impl<'a, W: Write + Seek, T: ColorType> ImageEncoder<'a, W, T> {
+    fn new(mut encoder: DirectoryEncoder<'a, W>, width: u32, height: u32) -> TiffResult<ImageEncoder<'a, W, T>> {
+        let row_samples = u64::from(width) * <T>::bits_per_sample().len() as u64;
+        let row_bytes = row_samples * u64::from(<T::Inner>::BYTE_LEN);
+
+        // As per tiff spec each strip should be about 8k long
+        let rows_per_strip = (8000 + row_bytes - 1) / row_bytes;
+
+        let strip_count = (u64::from(height) + rows_per_strip - 1) / rows_per_strip;
+
+        encoder.write_tag(Tag::ImageWidth, width);
+        encoder.write_tag(Tag::ImageLength, height);
+        encoder.write_tag(Tag::Compression, 1u16);
+
+        encoder.write_tag(Tag::BitsPerSample, &<T>::bits_per_sample() as &[u16]);
```

Comment:
Here is the superfluous allocation, `bits_per_sample() -> Vec<_>`. No cast, and passing the value by reference would shorten this to `write_tag(Tag::BitsPerSample, <T>::bits_per_sample);`